libbeat/reader/debug/debug.go

```diff
@@ -0,0 +1,178 @@
+	failures      int
+	predicate     CheckFunc
+	state         state
+	offset        int64
+}
+
+// NewReader returns a debug reader.
+func NewReader(
+	log *logp.Logger,
+	reader io.Reader,
+	minBufferSize int,
+	maxFailures int,
+	predicate CheckFunc,
+) (*Reader, error) {
+	return &Reader{
+		log:           log,
+		minBufferSize: minBufferSize,
+		reader:        reader,
+		maxFailures:   maxFailures,
+		predicate:     predicate,
+	}, nil
+}
+
+// Read will proxy the read call to the original reader and will periodically checks the values of
+// bytes in the buffer.
+func (r *Reader) Read(p []byte) (int, error) {
+	if r.state == stopped {
+		return r.reader.Read(p)
+	}
+
+	if r.state == running && r.failures > r.maxFailures {
+		// cleanup any remaining bytes in the buffer.
+		if r.buffer.Len() > 0 {
+			r.predicate(r.offset, r.buffer.Bytes())
+			r.buffer = bytes.Buffer{}
+		}
+		r.log.Info("Stopping debug reader, max execution reached")
+		r.state = stopped
+		return r.reader.Read(p)
+	}
+
+	if r.state == initial {
+		r.log.Infof(
+			"Starting debug reader with a buffer size of %d and max failures of %d",
+			r.minBufferSize,
+			r.maxFailures,
+		)
+		r.state = running
+	}
+
+	n, err := r.reader.Read(p)
+
+	if n != 0 {
+		r.buffer.Write(p[:n])
+		if r.buffer.Len() >= r.minBufferSize {
+			if r.failures < r.maxFailures && r.predicate(r.offset, r.buffer.Bytes()) {
+				r.failures++
+			}
+			r.buffer = bytes.Buffer{}
+		}
+		r.offset += int64(n)
+	}
+	return n, err
+}
+
+func makeNullCheck(log *logp.Logger, minSize int) CheckFunc {
+	// create a slice with null bytes to match on the buffer.
+	pattern := make([]byte, minSize, minSize)
+	return func(offset int64, buf []byte) bool {
+		idx := bytes.Index(buf, pattern)
+		if idx <= 0 {
+			offset += int64(len(buf))
+			return false
+		}
+		reportNull(log, offset+int64(idx), idx, buf)
+		return true
+	}
+}
+
+func reportNull(log *logp.Logger, offset int64, idx int, buf []byte) {
+	relativePos, surround := summarizeBufferInfo(idx, buf)
+	log.Debugf(
+		"Matching null byte found at offset %d (position %d in surrounding string: %s, bytes: %+v",
+		offset,
+		relativePos,
+		string(surround),
+		surround)
+}
+
+func summarizeBufferInfo(idx int, buf []byte) (int, []byte) {
+	startAt := idx - offsetStart
+	var relativePos int
+	if startAt < 0 {
+		startAt = 0
+		relativePos = idx
+	} else {
+		relativePos = offsetStart
+	}
+
+	endAt := idx + offsetEnd
+	if endAt >= len(buf) {
+		endAt = len(buf)
+	}
+	surround := buf[startAt:endAt]
+	return relativePos, surround
+}
+
+// AppendReaders look into the current enabled log selector and will add any debug reader that match
+// the selectors.
+func AppendReaders(reader io.Reader) (io.Reader, error) {
+	var err error
+
+	if logp.HasSelector("detect_null_bytes") || logp.HasSelector("*") {
+		log := logp.NewLogger("detect_null_bytes")
+		if reader, err = NewReader(
+			log,
+			reader,
+			defaultMinBuffer,
+			defaultMaxFailures,
+			makeNullCheck(log, 1),
```

Comment:
value of 1 is ok if encoding is utf-8 (and for a many other encodings). Storage is normally a block-device and one reads into bigger buffers -> we are very likely to see bursts of 0. We should set the minSize to at least 4, so to deal with more exotic encodings (e.g. utf-16/utf-32 or even more exotic ones). E.g. utf-16 is very standard on windows. With minSize of 1 this debugger will be of no use on most windows systems.